### PR TITLE
send and accept invitation

### PR DESCRIPTION
**Pull request by SunmadeBabafemi**
1. an invitation entity was created
2. a user can send an invite to another user's email to become a contributor to his api
3. the invitee will only have yo click the accept link in his email to accept the invite
4. the invitation only lasts for 30 days
5. after accepting the invite the corresponding model gets deleted from the database

### CODE REVIEW

**Comment by Chimexx** [comment]
Seen, I have some observations... I will review later today.

File: src/api/api.controller.ts [hunk 1]
```diff
@@ -150,4 +152,14 @@ export class ApiController {
     const apis = await this.apiService.getPopularAPis();
     return ZaLaResponse.Ok(apis, 'OK', '200');
   }
+
+  @Get('/contributors/:apiId')
+  @ApiOperation({summary: 'Get contributors of an API'})
+  async getApiContributors(
```
Comment (Chimexx) [inline src/api/api.controller.ts:158]: Add IdCheck guard here to verify if apiId exists in the db,
Also call the Authorization and Authentication guard if needed.

File: src/api/api.service.ts [hunk 2]
```diff
@@ -422,4 +422,38 @@ export class ApiService {
...
+      const api = await this.apiRepo.findOne({where:{id:apiId}})
+      if(!api){
+        throw new NotFoundException(
+          ZaLaResponse.NotFoundRequest(
+            "Not Found Error",
+            "Api Not Found",
+            "404"
+          )
+        )
+      }
```
Comment (Chimexx) [inline src/api/api.service.ts:438]: If you added IdCheck guard, there won't be need to check if api exists here or any other place.

File: src/api/api.service.ts [hunk 3]
```diff
@@ -422,4 +422,38 @@ export class ApiService {
...
+      const contributors = []
+      const contributorIds = api.contributors
+      for(const id of contributorIds){
+        const contributor = await this.profileRepo.findOneBy({id})
+        if(contributor){
+          contributors.push(contributor)
+        }
+      }
+      return contributors
```
Comment (Chimexx) [inline src/api/api.service.ts:447]: I would do this instead 
image

File: src/invitation/invitation.service.ts [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        )
+      }
+
+      const allInvites = await this.invitationRepo.find({where:{apiId: apiId}})
+     
+
+      return allInvites
+    } catch (error) {
+      console.log(error);  
+        throw new BadRequestException(
```
Comment (Chimexx) [inline src/invitation/invitation.service.ts:232]: remove logger

File: src/invitation/invitation.service.ts [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+
+      //Check if API exists
+      const api = await this.apiRepo.findOne({where:{id:apiId}})
+      if(!api){
+        throw new NotFoundException(
+          ZaLaResponse.NotFoundRequest(
+            "Not Found Error",
+            "Api Not Found",
+            "404"
+          )
```
Comment (Chimexx) [inline src/invitation/invitation.service.ts:222]: Again, Idcheck guard prevents this.

File: src/invitation/invitation.service.ts [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+      const verify = await this.jwTokenService.verify(existingInvite.token,{secret: invitationSecret})
+      contributors.push(verify.inviteeId)
+      
+      await this.apiRepo.update(
+        api.id,
+        {contributors: contributors},
+
+      )
+      //Delete the invitation after it has been accepted
+      await this.invitationRepo.delete(existingInvite.id)
```
Comment (Chimexx) [inline src/invitation/invitation.service.ts:186]: This should be synchronous

File: src/invitation/invitation.service.ts [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+      console.log(error);  
+      //  if (error.name === 'JsonWebTokenError') {
+      //   throw new UnauthorizedException(
+      //     ZaLaResponse.BadRequest(
+      //       'Subscription Error',
+      //       "Invitation already expired",
+      //       '403',
+      //     ),
+      //   );
+      // } else {
```
Comment (Chimexx) [inline src/invitation/invitation.service.ts:116]: Remove if not needed

**Review by Chimexx — CHANGES_REQUESTED** [review]
Nice, just a few more changes

File: src/invitation/invitation.service.ts (hunk 6 above)
Comment (SunmadeBabafemi) [reply src/invitation/invitation.service.ts:186]: thanks boss... i'll get to it

File: src/invitation/invitation.controller.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  async createInvite(
+    @Body() createInvitationDto: CreateInvitationDto,
+    @Param('apiId') apiId: string 
+    ): Promise<Ok<String>> {
+    const invite = await this.invitationService.createInvitation(createInvitationDto, apiId);
+    return ZaLaResponse.Ok(invite, 'Invite created', 201)
+  }
+
+  @Get('/accept/:apiId/:inviteeId')
+  @IdCheck('apiId')
```
Comment (johnayinde) [inline src/invitation/invitation.controller.ts:28]: You can use one Idcheck instead separated with commas

**Review by johnayinde — CHANGES_REQUESTED** [review]

**Review by Chimexx — COMMENTED** [review]
What of Authentication guard? Any one without a token can access these endpoints, unless you plan to make them public

**Comment by SunmadeBabafemi** [comment]
> What of Authentication guard? Any one without a token can access these endpoints, unless you plan to make them public

i just added guards to the send invite and view invitees.... for accepting the invite, the endpoint is hit from the invitee's email.. there is no guard that can be inserted there

**Pull request merged**